### PR TITLE
Add support for async lightbulb model.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Examples of this in action:

Note: the delays are exaggerated here to see the impact.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:231]: this is the API we think we'll be implementing from the platform.  However, this is still in teh air as to what the final shape will be.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:369]: the `bool?` is effectively a tri-state.  It means:

1. `true`: i want high pri, only give me high pri.
2. `false`: i want noral pri, only give me normal pri.
3. `null`: i don't care.  give me both. (legacy behavior).

An enum would likely be better here, i just wasted sure what teh right location was to define it.

File: src/Features/Core/Portable/CodeFixes/CodeFixService.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeFixes/CodeFixService.cs:402]: code fixing, refactoring, and analyzers all effectively have a way to say: if you want high-pri or normal-pri, i'll limit to that set.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:247]: @mavasani there's still something i'm uncertain about.  It appears as if diagnostics has both a push and pull mode.  I think i've update the pull side so taht if lightbulbs cause us to compute this data, that we compute the high pri items first, then the normal pri.

However, there's still the case that the incremental analyzer is running and producing (i.e pushing) results, and i'd like it to also run teh priority analyzers first in that case.

Am i widely off base here? Should that be something we want to do in this pr?  Or is it unnecessary here for reasons that i don't quite understand currently :)

Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:247]: You are correct. Your current code changes are fine, but only work for the Light bulb path for fetching diagnostics for a span. I think you want to move the code filtering the analyzers based on the priority further down the callstack, i.e. pass in `highPriority` flag to [DocumentAnalysisScope ctor](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/DocumentAnalysisScope.cs,b80ffac13be08af4,references). This scope will be created for both push/pull diagnostics - you just to ensure that you pass in the appropriate priority values at all construction sites for `DocumentAnalysisScope`

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:212]: Instead of filtering above, pass in the priority here and let `DocumentAnalysisExecutor` do the filtering. Basically, `executor.ComputeDiagnosticsAsync` will have an early bail out check for priority matching.

**Review by mavasani — COMMENTED** [review]
Overall, this looks great and matches the implementation I had in mind. Just moving the filtering a bit further down should make the behavior common across all IDE document diagnostic computation code paths.

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (alrz) [inline src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:18]: I think `T:` is unneccary and the suppression can be removed?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:18]: this is the workspaces layer talking about types in teh features layer.

File: src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs:148]: @mavasani  i'm assuming it's fine to report multiple sets of diagnostics for the same doc.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: @KirillOsenkov @Therzok  would you guys intend to support this in VS4mac?  If so, it would allow roslyn to remove a bunch of the legacy code for ISuggestedActionsSource3

Comment (KirillOsenkov) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: @davidwengier @sevoku

Comment (jasonmalinowski) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: Note this file is also linked into EditorFeatures.Cocoa so it'll apply to both unless we do something else here.

Comment (Therzok) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: What exactly does support this mean in this context?

Comment (davidwengier) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: > Note this file is also linked into EditorFeatures.Cocoa so it'll apply to both unless we do something else here.

Extension to this note: We're currently not inserting any bits from Roslyn main into VS Mac, so merging this could cause a lot of pain when we do start up again (once d17 vs mac has settled down), and cause a bunch of extra work at a really bad time.

Perhaps a better approach for the short term would be to un-link these files, and keep the old versions (or part thereof with as much code sharing as possible) in the Cocoa project.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:35]: i'm going to pull this new code into a new linked partial part.  we just won't include that in cocoa.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs]: Switched to an enum.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:114]: What is the WhenAll really doing here? You could delete this and it'd all work the same, right?

**Comment by ryzngard** [comment]
@CyrusNajmabadi  what's the behavior when no items have been calculated yet? In the examples it looks like at least 1 item is available and then a loading bar shows

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (jasonmalinowski) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:114]: (my assumption was either we'd use the result from WhenAll, or we'd do .Result on the tasks since we know they're then complete or something. It made me wonder if I had mis-read the code and the tasks on line 114 and on 115-116 were actually different.)

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:83]: I can't really speak to this much, but any concern that we're potentially running this twice, both for priority = high and priority = normal?

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:285]: No await needed?

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs
Comment (jasonmalinowski) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs:35]:
```suggestion
        /// Add-using gets special privileges as being the most used code-action, along with being a core
```

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:9]: Might be worth adding a comment that you're using this because you're referencing a higher layer.

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:12]:
```suggestion
    /// the highest precedence when ordering items so that we can ensure very important items get top prominence,
```

File: src/Workspaces/Core/Portable/CodeActions/CodeActionProviderPriority.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeActions/CodeActionProviderPriority.cs:11]: I admit the naming here is a bit confusing compared to CodeActionPriority; this is only really used to filter when we're computing, right? i.e. this isn't returned from CodeActionProvider anywhere. Maybe call it "CodeActionRequestedPriority" or something?

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:28]: Should we have a check that asserts that if we end up with a high-priority action that the provider was marked as high priority? I worry we might do the former but not the latter should we ever add a new one and end up with a flaky experience.

(It's also possible you did add this and I missed it...)

File: src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs
Comment (jasonmalinowski) [inline src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs:120]: Not urgent but I wonder if this is another one to mark. Honestly if I'm ever invoking a fix on the usings block, I more or less assume it's always this one too.

**Review by jasonmalinowski — APPROVED** [review]
Looks fine to me, although I admit I don't well understand our layers of how we process and filter diagnostics to know if the priority checks are in the "right" place. Somebody else (@mavasani or @sharwell, probably) should sign off on those too.

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:9]: Will do

File: src/Workspaces/Core/Portable/CodeActions/CodeActionProviderPriority.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeActions/CodeActionProviderPriority.cs:11]: I like that. Thanks!

File: src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeActions/CodeActionPriority.cs:28]: Seems reasonable!

File: src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs:120]: Sure. But I'll do that in a following pr. I'll open issue to track this.

File: src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer_Core.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer_Core.cs:47]: Yes. This should match the new priority type I added. Will fix this.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:54]: It's configureawait(true).

File: src/Features/Core/Portable/CodeFixes/CodeFixService.cs
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/CodeFixes/CodeFixService.cs:402]: Yes. I will clean this up. Good request.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:199]: I was refactoring this space but ended up undoing other stuff. But I left this as good hygiene :-)

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:44]: So this t mirrors what the sync code already does. So if prefer to just let things the same in that regard.

File: src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs:120]: Tracking this here: https://github.com/dotnet/roslyn/issues/53705

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs]: agreed.  fixed.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs]: simplified majorly (containing method is just removed).

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (KirillOsenkov) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: Back in the olden days we always required a block statement under control flow structures. Did this change or get relaxed?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: this is relaxed.  the rule is: if it's a single-line you don't need the block.

Comment (KirillOsenkov) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: Not going to argue, but in my opinion this is a regression that makes it worse. A few times we've been burned badly after a bad merge or just someone not paying attention. Much harder to make this mistake when curlies are present.

Of course I understand you're not going to change it back now :) Grumble grumble

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: > Much harder to make this mistake when curlies are present.

it should be very hard to make a mistake here.  For example, teh following will not compile:

```c#
if (foo)
    bar();
    bar();
```

This will fail as the formatting will be incorrect.  

--

Note: we haven't had a single issue with instituting this as we also have hte formatting checker.

Comment (KirillOsenkov) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: I've mostly learned, refined and solidified my coding style from you during the early Roslyn days, and our StyleCop choices in the old repo. If you look at the code I wrote in the past 10 years it'll be factored and formatted the way you'd do it.

So if you tell me to change it now, I guess I have no choice but adapt! Growth mindset and all.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: :)

i was resistent to this at first.  but we wanted to adopt the .net coding pattern, so this came along.  I found it actually somewaht reasonable.  The one thing i can't abide by is:

```c#
if (foo)
   bar();
else
   baz();
```

That just still looks awful to me :)

Importantly though, the block can be avoided *only* if the body is on a *single line*.  so this is allowed:

```c#
if (foo)
   bar(baz(), quux());
```

But this is not:

```c#
if (foo)
    bar(
        baz(), quux());
```

The restrictions make this look good, while avoiding lots of confusing cases.

Comment (davidwengier) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:116]: When I see formatting rules like this I just close my eyes and think of my side projects 😛

**Review by davidwengier — APPROVED** [review]

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:54]: > Is it worth specifying?

Sure.  I can do that :)

> Also do you need to specify .WithCancellation(..) too?

My understanding was 'no'.  But i'll check on that as i'mnot super familiar with IAsyncEnumerable.

File: src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs:121]: Effectively, everything that previously produced a 'high pri code action' is in the 'high pri request' category.  Anything that produces a normal/low pri code action is in the normal request category.  I didn't want to change anything in that regard here.  Right now we've had a long historical precedent of recognizing two code actions (one analzyer, one refactoring) that are so important that they should beat out everything else.  This PR codifies things so that not only do those come first, they are computed first so we can display them asap for the user.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:114]: yes.  very deliberate to still wait on this group.  Outside of hte high-pri group, we want the rest of hte actions to still be sorted in the same manner as today.  Because of that, we still need to wait until they're done and merge/sort the results before presenting htem.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:69]: sure!

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs
Comment (davidwengier) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs:54]: > i'mnot super familiar with IAsyncEnumerable.

You've now used it one more time than me, so you're the closest thing to an expert I know :P

**Pull request merged**
